USB: harness: Improve testability

finikorg opened the pull request:
Enable for `native_posix` `usb_device`. Switch on more tests, fix issues found in the tests.

Command to execute tests:
```
sanitycheck --device-testing --device-serial /dev/ttyACM0 -p <board> -t usb
```

finikorg committed:
samples: usb: hid: Enable harness for the sample

Enable harness so we can execute smoke tests on the hardware with
command line:
sanitycheck --device-testing --device-serial /dev/ttyACM0 -p <BOARD>
  -T samples/subsys/usb/hid

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: tests: Enable other boards for test

Enable other boards with USB device supported

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: usb: Add depends_on gpio

Add gpio dependence, fixing build for native_posix.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
tests: usb: Fix checking error code

Instead of checking for -EINVAL check that return is not successful
since USB device may return other error code.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
native_posix: Enable USB device for the board

Enable USB device and turn on appropriate tests in sanity check.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
tests: usb: Add usb_read() for invalid endpoint

Check that reading invalid endpoint does not succeed.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
tests: usb: Add usb_read() for invalid endpoint

Check that reading invalid endpoint does not succeed.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

carlescufi commented:
@niko @masz-nordic can you run this on nrf52840_pca10056?

aescolar reviewed (commented):
I'm ok with the native_posix changes.

finikorg committed:
samples: usb: mass: Add whitelist mass storage on ram

To support mass storage over USB in RAM we need a board with
sufficient RAM. Fix failing tests for boards with low memory.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
tests: usb: Cleanup tests

General test cleanup and remove optimized variables.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

masz-nordic commented:
nrf52840_pca10056 fails following tests:
- samples/subsys/usb/dfu/test - timeout
handler.log:
```
***** Booting Zephyr OS v1.14.0-rc1-1399-g0002598c309c *****

[00:00:00.000,000] <dbg> main.composite_pre_init: HID Device: dev 0x20003724

[00:00:00.000,000] <dbg> main.main: Starting application

[00:00:00.003,509] <dbg> main.status_cb: status 2 unhandled

[00:00:00.006,622] <dbg> main.status_cb: status 5 unhandled
```
- tests/subsys/usb/device/usb.device - timeout
handler.log is empty
- tests/subsys/usb/bos/usb.bos - timeout
handler.log is empty

finikorg committed:
tests: usb: Add write() and set_callback() tests

Add more tests for invalid EP testing.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
tests: usb: Add write() and set_callback() tests

Add more tests for invalid EP testing.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: usb_dc_dw: Fix valid endpoints check

Fix check for valid endpoints; issue is found in harness testing.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: usb_dc_native_posix: Fix valid endpoints check

Fix check for valid endpoints; issue is found in harness testing.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

carlescufi commented:
@finikorg thanks for this!

There are 2 failures with nRF52840. One is expected (DFU timeout) but the other one is not:

```
$ scripts/sanitycheck --device-testing --device-serial /dev/ttyACM0 -p nrf52840_pca10056 -t usb
JOBS: 16
Building testcase defconfigs...
12 tests selected, 92405 tests discarded due to filters
total complete:    4/  12  33%  failed:    0

nrf52840_pca10056         samples/subsys/usb/dfu/test                        FAILED: timeout
        see: sanity-out/nrf52840_pca10056/samples/subsys/usb/dfu/test/handler.log

total complete:    9/  12  75%  failed:    1

nrf52840_pca10056         tests/subsys/usb/device/usb.device                 FAILED: timeout
        see: sanity-out/nrf52840_pca10056/tests/subsys/usb/device/usb.device/handler.log

total complete:   12/  12  100%  failed:    2
10 of 12 tests passed with 0 warnings in 223 seconds
```

finikorg committed:
usb: usb_dc_dw: Check endpoint is valid

Fix harness test issues.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: usb_dc_native_posix: Check endpoint is valid

Fix harness tests.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: usb_dc_kinetis: Verify endpoint is valid

Verify endpoint before usb_dc_ep_set_callback().

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: usb_dc_kinetis: Fix using invalid index

I does make sense to use index only after we make sure it is valid,
issue is found in harness tests.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
tests: usb: Add more USB DC tests and refactor names

Add more tests covering almost all DC API with invalid parameters.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

masz-nordic commented:
I have tried rerunning tests for a few times - they seem very inconsistent. I get from 2 up to 6 failures depending on rerun.

finikorg committed:
usb: usb_dc_stm32: Add missing function

Add missing API functions which will be tested with harness tests.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

jfischer-phytec-iot commented on tests/subsys/usb/device/src/main.c:
```diff
@@ -158,7 +158,6 @@ static void test_usb_deconfig(void)
 /* Test USB Device Cotnroller API */
 static void test_usb_dc_api(void)
 {
-	/* Control endpoins are configured */
```
Something is wrong with this commit, it does not change any functionality in the code.

finikorg replied:
Fixed, somehow the patch was incorrectly rebased.
